src/scmdata/database.py

```diff
@@ -51,6 +52,248 @@ def _check_is_subdir(root, d):
 
 
+def _get_safe_filename(inp):
+    def safe_char(c):
+        if c.isalnum() or c in "-/*_.":
+            return c
+        else:
+            return "-"
+
+    return "".join(safe_char(c) for c in inp)
+
+
+class DatabaseBackend(ABC):
+    """
+    Abstract backend for serialising/deserialising data
+
+    Data is stored as objects represented by keys. These keys can be used later
+    to load data.
+    """
+
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+
+    @abstractmethod
+    def save(self, sr):
+        """
+        Save data
+
+        Parameters
+        ----------
+        sr: scmdata.ScmRun
+
+        Returns
+        -------
+        str
+            Key where the data is stored
+        """
+        pass
+
+    @abstractmethod
+    def load(self, key):
+        """
+        Load data at a given key
+
+        Parameters
+        ----------
+        key : str
+            Key to load
+
+        Returns
+        -------
+        scmdata.ScmRun
+        """
+        pass
+
+    def delete(self, key):
+        """
+        Delete a given key
+
+        Parameters
+        ----------
+        key: str
+        """
+        pass
+
+    @abstractmethod
+    def get(self, filters):
+        """
+        Get all matching keys for a given filter
+
+        Parameters
+        ----------
+        filters: dict of str
+            String filters
+            If a level is missing then all values are fetched
+
+        Returns
+        -------
+        list of str
+            Each item is a key which may contain data which is of interest
+        """
+        pass
+
+
+class NetCDFBackend(DatabaseBackend):
+    """
+    On-disk database handler for outputs from SCMs
+
+    Data is split into groups as specified by :attr:`levels`. This allows for fast
+    reading and writing of new subsets of data when a single output file is no longer
+    performant or data cannot all fit in memory.
+    """
+
+    def get_key(self, sr):
+        """
+        Get key where the data will be stored
+
+        The key is the root directory joined with the other information provided. The filepath
+        is also cleaned to remove spaces and special characters.
+
+        Parameters
+        ----------
+        sr : :class:`scmdata.ScmRun`
+            Data to save
+
+        Raises
+        ------
+        ValueError
+            If non-unique metadata or missing metadata is found for each of :attr:`self.kwargs["levels"]`
+
+        Returns
+        -------
+        str
+            Path in which to save the data without spaces or special characters
+        """
+        levels = {
+            level: sr.get_unique_meta(level, no_duplicates=True).replace(os.sep, "_")
+            for level in self.kwargs["levels"]
+        }
+
+        return self._get_out_filepath(**levels)
+
+    def _get_out_filepath(self, **levels):
+        out_levels = []
+        for level in self.kwargs["levels"]:
+            if level not in levels:
```

Comment:
which way around does this check? Does it check that the user supplied level is in the levels used to initialise the database (what I presume is `self.kwargs["levels"]`) or does it check that the levels used to initialise the database are all in the user supplied levels?

Reply:
It checks that the user's data has all of the levels  required by the database. so the later

Reply:
can we make that clearer with the naming?